api/lib/domain/usecases/create-user-and-reconcile-to-schooling-registration-from-external-user.js

```diff
@@ -59,6 +59,11 @@ module.exports = async function createUserAndReconcileToSchoolingRegistrationFro
     const student = await studentRepository.getReconciledStudentByNationalStudentId(matchedSchoolingRegistration.nationalStudentId);
     await userReconciliationService.checkIfStudentHasAlreadyAccountsReconciledInOtherOrganizations(student, userRepository, obfuscationService);
 
+    const user = await userRepository.getBySamlId(externalUser.samlId);
+    if (user) {
+      return user;
+    }
+
```

Comment:
Si j'ai bien compris le fonctionnement de Knex.transaction (il ouvre une nouvelle transaction) et le mécanisme de gestion de concurrence par défaut [read commited](https://pgdash.io/blog/postgres-transactions.html), tester l'existence dans le use-case résoudra une partie des problèmes, mais pas tous.
> the second transaction could not read the first transaction’s as-yet-uncommitted data

Il peut toujours y avoir 2 appels parallèles à userRepository.`createAndReconcileUserToSchoolingRegistration` :
- 2 transactions T1 et T2 (qui font INSERT INTO users + UPDATE schooling_registrations) 
- T2 COMMIT la première => l'API renvoie OK
- T1 essaye de COMMIT après T2, l'erreur CONSTRAINT_VIOLATED est levée => l'API rnevoie 500

Je pense que la solution est de catcher l'erreur (spécifiquement) dans le repo. Après, soit tu relis la table users dans le repo, soit tu laisses le use-case le faire.

Un exemple tiré de `knowledge-element-snapshot-repository`
```
catch (error) {
      if (bookshelfUtils.isUniqConstraintViolated(error)) {
        throw new AlreadyExistingEntity(`A snapshot already exists for the user ${userId} at the datetime ${snappedAt}.`);
      }
```

Je veux bien l'avis de quelqu'un de référence BDD (@jonathanperret ou @laura-bergoens ?) pour confirmer ça.

Reply:
Oui, il peut rester des cas où une transaction n'arrive pas à terme à cause d'un conflit avec une autre. Il faut remarquer cependant que l'échelle de temps dont on parle est différente du cas qui est corrigé ici. Ici, on vérifie l'absence de compte utilisateur, on présente un formulaire à l'utilisateur, puis au retour de ce formulaire on faisait l'hypothèse que le compte utilisateur était toujours inexistant. L'intervalle de temps entre la vérification et l'insertion était pratiquement illimité, ce qui explique la fréquence des conflits observés.

Pour le cas où le test est "très proche" (notamment, sans interaction humaine) de l'insertion, le risque de conflit est bien plus faible. Il reste possible en général de le déclencher, typiquement en bombardant l'API de requêtes similaires, mais a priori pas avec l'interface utilisateur (où on met généralement un verrouillage côté client pour éviter les clics à répétition).

Le cas de conflit devient donc extrêmement rare (jusqu'à présent, tous les cas dans Pix qui ont été traités de la façon que je décris ci-dessus n'ont plus donné lieu à des conflits). À partir de là, mon opinion est qu'écrire une gestion de conflits sous la forme d'un gestionnaire d'exception est quelque peu futile. En effet, ce gestionnaire d'exception sera nécessairement fragile : dans l'exemple que tu donnes, le test `bookshelfUtils.isUniqConstraintViolated(error)` ne permet pas vraiment d'affirmer ce que dit le message d'erreur (`A snapshot already exists for the user…`) car on ne regarde même pas quelle contrainte a été violée. Bien sûr on peut avoir regardé la table en question et vérifié qu'une seule contrainte d'unicité est en jeu, mais cela peut évoluer dans le temps. De plus, le gestionnaire d'exception n'est pas spécifique à une requête ! A priori, si on a ouvert une transaction c'est pour effectuer plusieurs requêtes, comment affirmer laquelle est à la source de l'exception ? Il existe des solutions à tous ces points mais une implémentation "complète" commencerait à ressembler à une usine à gaz.

De plus, une fois écrit le gestionnaire d'exception il est tentant de s'en servir pour gérer les cas "quasi-nominaux" du type : essayer d'insérer un utilisateur avec un adresse e-mail existant déjà. Après tout si on est capable de gérer le conflit, pourquoi ne pas insérer l'adresse saisie et gérer ce qui se passe (au lieu de vérifier préalablement) ? Si on prenait cette approche, à tous les problèmes mentionnés plus haut s'ajouterait celui du _monitoring_ : la base de données rencontrerait des requêtes en "erreur" à longueur de journée, rendant d'autant plus difficile de faire le tri pour identifier les vrais problèmes.

Il existe aussi des approches intermédiaires, utilisant par exemple du "verrouillage optimiste" (conditionner l'instruction d'insertion de sorte qu'elle n'insère rien en cas de doublon) ou encore du verrouillage explicite (type `SELECT … FOR UPDATE`), applicables ou non selon les cas. Mais pour ma part je m'attends à ce que l'ajout proposé par @MathieuGilet soit adéquat pour prévenir les conflits qui pourront se présenter de manière réaliste.

Reply:
merci @jonathanperret. Je n'avais pas compris le scéanrioi d'usage
> Ici, on vérifie l'absence de compte utilisateur, on présente un formulaire à l'utilisateur, puis au retour de ce formulaire on faisait l'hypothèse que le compte utilisateur était toujours inexistant. L'intervalle de temps entre la vérification et l'insertion était pratiquement illimité, ce qui explique la fréquence des conflits observés.

C'est bon pour moi en l'état !